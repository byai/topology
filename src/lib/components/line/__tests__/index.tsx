/* eslint-disable no-undef */
import React from 'react';
import { cleanup, render, fireEvent } from '@testing-library/react';
import 'jest-dom/extend-expect';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import Line from '..';

expect.extend({ toMatchDiffSnapshot });
afterEach(cleanup);

describe('line', () => {
    test('正常render', () => {
        const div = document.createElement('div');
        const { baseElement } = render(
            <svg>
                <Line
                    start={{ x: 0, y: 0 }}
                    end={{ x: 100, y: 100 }}
                />
            </svg>,
            {
                container: document.body.appendChild(div),
            },
        );
        expect(baseElement).toMatchSnapshot();
    });
    test('选中时', () => {
        const div = document.createElement('div');
        const { baseElement } = render(
            <svg>
                <Line
                    start={{ x: 0, y: 0 }}
                    end={{ x: 100, y: 100 }}
                    selected
                />
            </svg>,
            {
                container: document.body.appendChild(div),
            },
        );
        expect(baseElement).toMatchSnapshot();
    });
    test('有数据时', () => {
        const div = document.createElement('div');
        const { baseElement } = render(
            <svg>
                <Line
                    start={{ x: 0, y: 0 }}
                    end={{ x: 100, y: 100 }}
                    selected
                    data={{ start: '1-0', end: '2' }}
                />
            </svg>,
            {
                container: document.body.appendChild(div),
            },
        );
        expect(baseElement).toMatchSnapshot();
    });
    test('鼠标操作： 第一个path移入、移出、点击', () => {
        const div = document.createElement('div');
        const { asFragment, container } = render(
            <svg>
                <Line
                    start={{ x: 0, y: 0 }}
                    end={{ x: 100, y: 100 }}
                    data={{ start: '1-0', end: '2' }}
                />
            </svg>,
            {
                container: document.body.appendChild(div),
            },
        );
        const firstRender = asFragment();
        // 移入
        fireEvent.mouseEnter(container.firstElementChild);
        expect(firstRender).toMatchDiffSnapshot(asFragment());
        // 移出
        const secondRender = asFragment();
        fireEvent.mouseLeave(container.firstElementChild);
        expect(secondRender).toMatchDiffSnapshot(asFragment());
        // 点击
        fireEvent.click(container.firstElementChild);
        expect(firstRender).toMatchDiffSnapshot(asFragment());
    });

    test('鼠标操作： 第二个path移入、移出、点击', () => {
        const div = document.createElement('div');
        const { asFragment, container } = render(
            <svg>
                <Line
                    start={{ x: 0, y: 0 }}
                    end={{ x: 100, y: 100 }}
                    data={{ start: '1-0', end: '2' }}
                />
            </svg>,
            {
                container: document.body.appendChild(div),
            },
        );
        const firstRender = asFragment();
        // 移入
        fireEvent.mouseEnter(container.lastElementChild);
        expect(firstRender).toMatchDiffSnapshot(asFragment());
        // 移出
        const secondRender = asFragment();
        fireEvent.mouseLeave(container.lastElementChild);
        expect(secondRender).toMatchDiffSnapshot(asFragment());
    });
    test('单选、多选', () => {
        const div = document.createElement('div');
        const { container } = render(
            <svg>
                <Line
                    start={{ x: 0, y: 0 }}
                    end={{ x: 100, y: 100 }}
                    data={{ start: '1-0', end: '2' }}
                    onSelect={() => null}
                />
            </svg>,
            {
                container: document.body.appendChild(div),
            },
        );
        fireEvent.click(container.firstElementChild);
        fireEvent(
            container.firstElementChild,
            new MouseEvent('click', {
                metaKey: true,
                ctrlKey: true,
            }),
        );
    });
    test('取消选中', () => {
        const div = document.createElement('div');
        const { container } = render(
            <svg>
                <Line
                    start={{ x: 0, y: 0 }}
                    end={{ x: 100, y: 100 }}
                    selected
                    data={{ start: '1-0', end: '2' }}
                    onSelect={() => null}
                />
            </svg>,
            {
                container: document.body.appendChild(div),
            },
        );
        fireEvent.click(container.firstElementChild);
        fireEvent(
            container.firstElementChild,
            new MouseEvent('click', {
                metaKey: true,
                ctrlKey: true,
            }),
        );
    });
});
