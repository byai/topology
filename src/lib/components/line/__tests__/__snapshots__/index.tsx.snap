// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`line 有数据时 1`] = `
<div>
  <svg>
    <path
      d="
        M 0 0
        C 8, 30 92, 60 100 90
    "
      fill="none"
      stroke="transparent"
      stroke-width="20"
      style="pointer-events: all; transition: .2s;"
    />
    <path
      d="
        M 0 0
        C 8, 30 92, 60 100 90
    "
      fill="none"
      stroke="#1F8CEC"
      stroke-width="2.5"
      style="pointer-events: all; transition: .2s;"
    />
    <path
      class="byai-topology-line-end-triangle"
      d="
    M 100 90
    l 5 0
    l -5 10
    l -5 -10
    Z
"
      data-json="{\\"origin\\":{\\"start\\":\\"1-0\\",\\"end\\":\\"2\\"},\\"po\\":{\\"start\\":{\\"x\\":0,\\"y\\":0},\\"end\\":{\\"x\\":100,\\"y\\":100}}}"
      data-type="end"
      fill="#1F8CEC"
      stroke="none"
      style="pointer-events: all; transition: .2s;"
    />
  </svg>
</div>
`;

exports[`line 正常render 1`] = `
<div>
  <svg>
    <path
      d="
        M 0 0
        C 8, 30 92, 60 100 90
    "
      fill="none"
      stroke="transparent"
      stroke-width="20"
      style="pointer-events: all; transition: .2s;"
    />
    <path
      d="
        M 0 0
        C 8, 30 92, 60 100 90
    "
      fill="none"
      stroke="#AAB7C4"
      stroke-width="2.5"
      style="pointer-events: all; transition: .2s;"
    />
    <path
      class="byai-topology-line-end-triangle"
      d="
    M 100 90
    l 5 0
    l -5 10
    l -5 -10
    Z
"
      data-json=""
      data-type="end"
      fill="#AAB7C4"
      stroke="none"
      style="pointer-events: all; transition: .2s;"
    />
  </svg>
</div>
`;

exports[`line 选中时 1`] = `
<div>
  <svg>
    <path
      d="
        M 0 0
        C 8, 30 92, 60 100 90
    "
      fill="none"
      stroke="transparent"
      stroke-width="20"
      style="pointer-events: all; transition: .2s;"
    />
    <path
      d="
        M 0 0
        C 8, 30 92, 60 100 90
    "
      fill="none"
      stroke="#1F8CEC"
      stroke-width="2.5"
      style="pointer-events: all; transition: .2s;"
    />
    <path
      class="byai-topology-line-end-triangle"
      d="
    M 100 90
    l 5 0
    l -5 10
    l -5 -10
    Z
"
      data-json=""
      data-type="end"
      fill="#1F8CEC"
      stroke="none"
      style="pointer-events: all; transition: .2s;"
    />
  </svg>
</div>
`;

exports[`line 鼠标操作： 第一个path移入、移出、点击 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`line 鼠标操作： 第一个path移入、移出、点击 2`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`line 鼠标操作： 第一个path移入、移出、点击 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`line 鼠标操作： 第二个path移入、移出、点击 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`line 鼠标操作： 第二个path移入、移出、点击 2`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;
